src/cauchy.jl

```diff
@@ -1,54 +1,72 @@
 #### Cauchy matrix
 
 export Cauchy
+
 """
-[`Cauchy` matrix](http://en.wikipedia.org/wiki/Cauchy_matrix)
+    [`Cauchy` matrix](http://en.wikipedia.org/wiki/Cauchy_matrix)
+
+* `Cauchy(x,y)[i,j] = 1 / (x[i] + y[j])`
+* `Cauchy(x) = Cauchy(x,x)`
+* `Cauchy(k::Int) = Cauchy(1:k)`
+
+Both `x` and `y` can be any iterable (typically vectors),
+but all elements of `x` must have the same type; likewise for `y`.
+
 ```julia
-Cauchy(x,y)[i,j]=1/(x[i]+y[j])
-Cauchy(x)=Cauchy(x,x)
-cauchy(k::Number)=Cauchy(collect(1:k))
-
-julia> Cauchy([1,2,3],[3,4,5])
-3x3 Cauchy{Int64}:
- 0.25      0.2       0.166667
- 0.2       0.166667  0.142857
- 0.166667  0.142857  0.125
-
-julia> Cauchy([1,2,3])
-3x3 Cauchy{Int64}:
- 0.5       0.333333  0.25
- 0.333333  0.25      0.2
- 0.25      0.2       0.166667
-
-julia> Cauchy(pi)
-3x3 Cauchy{Float64}:
- 0.5       0.333333  0.25
- 0.333333  0.25      0.2
- 0.25      0.2       0.166667
+julia> Cauchy([2.0 1], (0, 1, 2))
+2×3 Cauchy{Float64, Matrix{Float64}, Tuple{Int64, Int64, Int64}}:
+ 0.5  0.333333  0.25
+ 1.0  0.5       0.333333
+
+julia> Cauchy(1:3)
+3×3 Cauchy{Rational{Int64}, UnitRange{Int64}, UnitRange{Int64}}:
+ 1//2  1//3  1//4
+ 1//3  1//4  1//5
+ 1//4  1//5  1//6
+
+julia> Cauchy(3)
+3×3 Cauchy{Rational{Int64}, UnitRange{Int64}, UnitRange{Int64}}:
+ 1//2  1//3  1//4
+ 1//3  1//4  1//5
+ 1//4  1//5  1//6
 ```
 """
-struct Cauchy{T} <: AbstractMatrix{T}
-    x::Vector{T} #
-    y::Vector{T} #
-end # immutable
-
-function Cauchy(k::Number)
-         Cauchy(collect(1:k),collect(1:k))
-end
+struct Cauchy{T,X,Y} <: AbstractMatrix{T}
+    x::X
+    y::Y
 
-function Cauchy(x::Vector)
-         Cauchy(x,x)
+    # The element type T corresponds to the reciprocal
+    # of the sum of pairs of elements of x and y
+    function Cauchy(x::X, y::Y) where {X, Y}
+        Tx = eltype(first(x))
+        Ty = eltype(first(y))
+        all(==(Tx), eltype.(x)) || throw(ArgumentError("inconsistent x element types"))
+        all(==(Ty), eltype.(y)) || throw(ArgumentError("inconsistent y element types"))
+        T = eltype(one(Tx) + one(Ty)) 
```

Comment:
Could we use the type promotion system here directly?

Reply:
Originally I had code that used something like `promote_type(x...)` but it failed `@inferred` for some tests. Apparently the compiler was unable to do type inference when the input `x` is an iterable with a mix of several element types.
So I resorted to this `all` approach to make type inference work.
I'm open to suggestions.  (Ideally Julia would have an abstract type for an iterable where all elements are the same type, and AbstractArray would be a subtype of that parent type.)